Admin/Remove-DuplicateEntriesFromIanaMappings.ps1

```diff
@@ -0,0 +1,205 @@
+        }
+
+        Write-Host "[+] Validating IanaTimeZoneMappings.xml on server $env:COMPUTERNAME"
+
+        $activityBase = "[$env:COMPUTERNAME]"
+        $writeProgressParams = @{
+            Activity = "$activityBase Getting IanaTimeZoneMappings.xml Path"
+            Id       = [Math]::Abs(($env:COMPUTERNAME).GetHashCode())
+        }
+        Write-Progress @writeProgressParams
+
+        # Locate the path where the IanaTimeZoneMappings.xml resides - to do this, read the MsiInstallPath from the registry
+        $exchangeServerSetupPathParams = @{
+            MachineName = $env:COMPUTERNAME
+            SubKey      = "SOFTWARE\Microsoft\ExchangeServer\v15\Setup"
+            GetValue    = "MsiInstallPath"
+        }
+        $exchangeSetupPath = Get-RemoteRegistryValue @exchangeServerSetupPathParams
+
+        if (([System.String]::IsNullOrEmpty($exchangeSetupPath))) {
+            Write-Host "[+] Unable to locate the Exchange Server setup path"
+            return
+        }
+
+        # Define the final full path to the mapping file and for the backup file (in case we need to create it)
+        $mappingFilePath = "$exchangeSetupPath\Bin\IanaTimeZoneMappings.xml"
+        $mappingBackupFilePath = "$mappingFilePath.{0}.bak" -f $(Get-Date -Format MMddyyyyHHmmss)
+
+        if ((Test-Path -Path $mappingFilePath) -eq $false) {
+            Write-Host "[+] Iana mappings file was not found" -ForegroundColor Red
+            Write-Progress @writeProgressParams -Completed
+
+            return
+        }
+
+        $writeProgressParams.Activity = $activityBase + " Searching for duplicate entries in IanaTimeZoneMappings.xml"
+        Write-Progress @writeProgressParams
+
+        # Check if IanaTimeZoneMappings.xml contains duplicate entries - this is done by the custom Test-IanaTimeZoneMapping function
+        $testIanaTimeZoneMappingResults = Test-IanaTimeZoneMapping -FilePath $mappingFilePath
+
+        if ($null -ne $testIanaTimeZoneMappingResults -and
+            $testIanaTimeZoneMappingResults.DuplicateEntries.Count -ge 1) {
+
+            $duplicateEntries = $testIanaTimeZoneMappingResults.DuplicateEntries
+            $ianaMappingXml = $testIanaTimeZoneMappingResults.IanaMappingXml
+
+            Write-Host "[+] Duplicate entries detected!" -ForegroundColor Yellow
+
+            try {
+                $writeProgressParams.Activity = $activityBase + " Creating backup $mappingBackupFilePath"
+                Write-Progress @writeProgressParams
+
+                # If duplicate entries were detected, create a backup of the file before modifying it
+                Copy-Item -Path $mappingFilePath -Destination $mappingBackupFilePath
+                Write-Host "[+] Backup created: $mappingBackupFilePath"
+            } catch {
+                Write-Host "[+] Failed to create backup file. Inner Exception: $_" -ForegroundColor Red
+                Write-Progress @writeProgressParams -Completed
```

Comment:
Recommend wrapping everything in a `try` `finally` block that handles the `Write-Progress` inside the `finally` block you will always do `Write-Progress @writeProgressParams -Completed` so you would just need to do a `return` when you don't want to continue with the code.

Reply:
Good idea. Will add this with the next iteration.